### DISCUSSION TITLE
Deploy Orchard Core with Docker Image

### DISCUSSION THREAD

**Discussion by sddev-dotnet** [OrchardCMS/OrchardCore#9214] [Q&A]
I am trying to deploy an orchard core headless cms installation in a docker container to my kubernetes cluster. I need to deploy it with the app installed, rather than having to configure it when the container starts up. So what I have done is connected to my remote database from my machine, started up and configured my orchard application, and created a Dockerfile that copies the app-data from my machine into the container. When I run the container locally, it works just fine. However, when I try and run it on my k8 cluster, I get the following error: 

![image](https://user-images.githubusercontent.com/70717046/115129408-757f7780-9f9a-11eb-83b9-8cccc48a9134.png)

I am not sure what could be causing the issue. 

Here is the Dockerfile
```
FROM mcr.microsoft.com/dotnet/sdk:3.1 as build


WORKDIR /src
COPY ./ ./

# Restore Nuget packages
RUN dotnet restore --configfile ./Nuget.config ./SDDev.Net.Admin.sln 


RUN dotnet publish -c Release -o /publish

COPY ./devops/App_Data ./publish/



#Publish stage
# To enable ssh & remote debugging on app service change the base image to the one below
# FROM mcr.microsoft.com/azure-functions/dotnet:2.0-appservice 
FROM mcr.microsoft.com/dotnet/aspnet:3.1

RUN apt-get update \
    && apt-get install -y --allow-unauthenticated \
        libc6-dev \
        libgdiplus \
        libx11-dev \
     && rm -rf /var/lib/apt/lists/*

WORKDIR /app
COPY --from=build /publish  ./


ENTRYPOINT ["dotnet", "SDDev.Net.Admin.dll"]
```

In the App_Data folder I am copying over it has: 
Sites
    Default
         appsettings.json
         ContentDefinition.json

**Comment by Skrypt** [comment]
This is a known issue from a recent patch made on .NET Core 3.1 which I believe has been fixed in the latest dev branch.
You need to upgrade OC to the latest nuget packages or merge latest changes from the dev branch if you are using source code.

**Reply by jtkech** [reply]
Yes, it is an out of range but not thrown by the mvc view buffer, here it is when we create a shell container

Here we clone app services group by group, for a given group we have a service descriptors collection and we resolve them on the fly to get an instances collection, so here it is like a service can't be resolved and we get less instances than descriptors OR like the app services are mutated while we are cloning them, but normally at this point the app startup should have completed, and no further processes should mutate the app services. This is the first time i see this issue.

**Reply by sddev-dotnet** [reply]
@jtkech is there anything I can share that would help figure out what is causing this? My site is all test data now, so I am fine sharing anything you need.

**Reply by sddev-dotnet** [reply]
I am going to try updating the nuget packages to the latest version tonight and see if that makes a difference.

**Comment by deanmarcussen** [comment]
It is more normal to mount the `App_Data` volume external to the container.

If you copy it in like this, then the data is going to be replaced every time you deploy a new container. so the cms will be less updatable.

https://docs.orchardcore.net/en/dev/docs/reference/core/Configuration/#orchard_app_data-environment-variable

It may or may not be related to your issue, but you might want to try it without copying the `App_Data` files in, and pointing to the existing folder (i.e. I assume you have it running locally, of that `App_Data` volume, through your environment variables, to see if you can get a working install.

**Reply by sddev-dotnet** [reply]
Thanks, I considered mounting a volume, but I thought the only thing in AppData was the connection string info, and since that won't be changing, I thought embedding it wouldn't be a problem. I was able to get the Azure Blob Service Shell working to where App_Data was pushed to blob storage, but that didn't get the container running
